File: frontend/src/types/matrix.ts
export type AssignmentType = "regular" | "support" | "emergency" | "self_planned";
export type AbsenceType = "vacation" | "sick" | "school" | "free" | "other";
export type SiteStatus = "active" | "paused" | "planned" | "completed" | "deleted";
export type MatrixCellMark = "orange" | "red" | "blue";

export type MatrixDay = {
  date: string;
  weekday: number;
  is_weekend: boolean;
};

export type MatrixPerson = {
  id: number;
  display_name: string;
  short_code: string;
  person_type: "internal" | "external" | "external_temp";
};

export type MatrixAssignment = {
  id: number;
  person: MatrixPerson;
  start_date: string;
  end_date: string;
  assignment_type: AssignmentType;
  note: string | null;
};

export type AssignmentRead = {
  id: number;
  site_id: number;
  person_id: number;
  start_date: string;
  end_date: string;
  assignment_type: AssignmentType;
  note: string | null;
  created_by_user_id: number | null;
  updated_by_user_id: number | null;
  created_at: string;
  updated_at: string;
};

export type MatrixAbsence = {
  person: MatrixPerson;
  absence_type: AbsenceType;
  start_date: string;
  end_date: string;
  note: string | null;
};

export type MatrixCell = {
  date: string;
  assignments: MatrixAssignment[];
  absences: MatrixAbsence[];
  mark: MatrixCellMark | null;
  conflict_level: "none" | "warning" | "hard" | string;
  conflict_reason: string | null;
  conflict_codes: string[];
};

export type MatrixSite = {
  id: number;
  site_number: string | null;
  name: string;
  location: string | null;
  customer: string | null;
  project_manager_person_id: number | null;
  project_manager: MatrixPerson | null;
  status: SiteStatus;
  info: string | null;
  color: string | null;
};

export type MatrixRow = {
  site: MatrixSite;
  cells: MatrixCell[];
};

export type MatrixResponse = {
  start_date: string;
  end_date: string;
  days: MatrixDay[];
  project_managers: MatrixPerson[];
  rows: MatrixRow[];
};

export type MatrixVersionResponse = {
  version: string;
  latest_updated_at: string | null;
};

export type MatrixEntryInput = {
  person_id?: number;
  external_name?: string;
};

export type MatrixConflictMessage = {
  severity: string;
  code: string;
  message: string;
  date: string | null;
};

export type MatrixMutationResponse = {
  warnings: MatrixConflictMessage[];
  infos: MatrixConflictMessage[];
  updated_cells: MatrixCell[];
};

File: backend/app/schemas/matrix.py
from __future__ import annotations

from datetime import date as Date

from pydantic import BaseModel, Field

from app.models.enums import AbsenceType, AssignmentType, MatrixCellMark, PersonType, SiteStatus


class MatrixDay(BaseModel):
    date: Date
    weekday: int
    is_weekend: bool


class MatrixPerson(BaseModel):
    id: int
    display_name: str
    short_code: str
    person_type: PersonType


class MatrixAssignment(BaseModel):
    id: int
    person: MatrixPerson
    start_date: Date
    end_date: Date
    assignment_type: AssignmentType
    note: str | None = None


class MatrixAbsence(BaseModel):
    person: MatrixPerson
    absence_type: AbsenceType
    start_date: Date
    end_date: Date
    note: str | None = None


class MatrixCell(BaseModel):
    date: Date
    assignments: list[MatrixAssignment]
    absences: list[MatrixAbsence]
    mark: MatrixCellMark | None = None
    conflict_level: str = "none"
    conflict_reason: str | None = None
    conflict_codes: list[str] = Field(default_factory=list)


class MatrixSite(BaseModel):
    id: int
    site_number: str | None = None
    name: str
    location: str | None = None
    customer: str | None = None
    project_manager_person_id: int | None = None
    project_manager: MatrixPerson | None = None
    status: SiteStatus
    info: str | None = None
    color: str | None = None

    model_config = {"from_attributes": True}


class MatrixRow(BaseModel):
    site: MatrixSite
    cells: list[MatrixCell]


class MatrixResponse(BaseModel):
    start_date: Date
    end_date: Date
    days: list[MatrixDay]
    project_managers: list[MatrixPerson] = Field(default_factory=list)
    rows: list[MatrixRow]


class MatrixEntryInput(BaseModel):
    person_id: int | None = None
    external_name: str | None = None


class MatrixCellPatch(BaseModel):
    site_id: int
    date: Date
    entries: list[MatrixEntryInput]


class MatrixRangePatch(BaseModel):
    site_id: int
    start_date: Date
    end_date: Date
    entries: list[MatrixEntryInput]


class MatrixCellMarkPatch(BaseModel):
    site_id: int
    date: Date
    mark: MatrixCellMark | None = None


class MatrixConflictMessage(BaseModel):
    severity: str
    code: str
    message: str
    date: Date | None = None


class MatrixMutationResponse(BaseModel):
    warnings: list[MatrixConflictMessage] = Field(default_factory=list)
    infos: list[MatrixConflictMessage] = Field(default_factory=list)
    updated_cells: list[MatrixCell] = Field(default_factory=list)
